commit 350b9a44089843ce2cc1d31e8c87894667f0f17b
Author: Orlando Vazquez <ovazquez@gmail.com>
Date:   2018-02-23T12:54:25-08:00 (1 year, 7 months ago)
    
    AGENT-1089 Hermes should have option to delete files matching a pattern and not upload them to Manta

diff --git a/actor/lib/findstream.js b/actor/lib/findstream.js
index 74e8804..e9cbbf1 100644
--- a/actor/lib/findstream.js
+++ b/actor/lib/findstream.js
@@ -8,55 +8,15 @@
  * Copyright (c) 2014, Joyent, Inc.
  */
 
-var mod_util = require('util');
-var mod_stream = require('stream');
-var mod_fs = require('fs');
 var mod_child = require('child_process');
 var mod_lstream = require('lstream');
-
+var mod_stream = require('stream');
+var mod_util = require('util');
 var mod_verror = require('verror');
 
-var FIND = '/usr/bin/find';
-
-function
-StatStream()
-{
-	mod_stream.Transform.call(this, {
-		objectMode: true
-	});
-}
-mod_util.inherits(StatStream, mod_stream.Transform);
+var mod_statstream = require('./statstream');
 
-StatStream.prototype._transform = function
-_transform(chunk, _, done)
-{
-	var self = this;
-
-	mod_fs.lstat(chunk, function (err, stat) {
-		if (err) {
-			if (err.code === 'ENOENT') {
-				/*
-				 * We probably raced with a file
-				 * removal.
-				 */
-				done();
-			} else {
-				done(err);
-			}
-			return;
-		}
-
-		if (stat.isFile()) {
-			self.push({
-				path: chunk,
-				mtime: stat.mtime,
-				size: stat.size
-			});
-		}
-
-		done();
-	});
-};
+var FIND = '/usr/bin/find';
 
 function
 FindStream(dirs)
@@ -67,7 +27,7 @@ FindStream(dirs)
 	});
 
 	self.ds_lstream = new mod_lstream();
-	self.ds_statstream = new StatStream();
+	self.ds_statstream = new mod_statstream.StatStream();
 
 	self.ds_stderr = '';
 	self.ds_destroyed = true;
diff --git a/actor/lib/listglob.js b/actor/lib/listglob.js
new file mode 100644
index 0000000..81de470
--- /dev/null
+++ b/actor/lib/listglob.js
@@ -0,0 +1,53 @@
+/*
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+/*
+ * Copyright (c) 2018, Joyent, Inc.
+ */
+
+var mod_child = require('child_process');
+var mod_statstream = require('./statstream');
+var mod_verror = require('verror');
+var mod_assert = require('assert-plus');
+
+var LS = '/usr/bin/ls';
+
+function listGlob(pattern, callback) {
+    mod_assert.string(pattern, 'pattern');
+    mod_assert.func(callback, 'callback');
+
+	var args = [
+		'-d'
+	];
+
+    args.push(pattern);
+
+	var opts = {
+		env: {},
+		cwd: '/'
+	};
+
+    var cmd = [LS, '-d', pattern].join(" ");
+
+	mod_child.exec(cmd, opts, onExec);
+
+    function onExec(err, stdout, stderr) {
+        if (err) {
+            callback(
+                new mod_verror.VError(err,
+                    'could not list files matching pattern'));
+            return;
+        }
+        callback(null, stdout.toString().trim().split('\n'));
+    }
+}
+
+
+module.exports = {
+    listGlob: listGlob
+};
+
+/* vim: set ts=8 sts=8 sw=8 noet: */
diff --git a/actor/lib/statstream.js b/actor/lib/statstream.js
new file mode 100644
index 0000000..942c305
--- /dev/null
+++ b/actor/lib/statstream.js
@@ -0,0 +1,58 @@
+/*
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+/*
+ * Copyright (c) 2018, Joyent, Inc.
+ */
+
+var mod_util = require('util');
+var mod_fs = require('fs');
+var mod_stream = require('stream');
+
+function StatStream()
+{
+	mod_stream.Transform.call(this, {
+		objectMode: true
+	});
+}
+mod_util.inherits(StatStream, mod_stream.Transform);
+
+StatStream.prototype._transform = function
+_transform(chunk, _, done)
+{
+	var self = this;
+
+	mod_fs.lstat(chunk, function (err, stat) {
+		if (err) {
+			if (err.code === 'ENOENT') {
+				/*
+				 * We probably raced with a file
+				 * removal.
+				 */
+				done();
+			} else {
+				done(err);
+			}
+			return;
+		}
+
+		if (stat.isFile()) {
+			self.push({
+				path: chunk,
+				mtime: stat.mtime,
+				size: stat.size
+			});
+		}
+
+		done();
+	});
+};
+
+module.exports = {
+	StatStream: StatStream
+};
+
+/* vim: set ts=8 sts=8 sw=8 noet: */
diff --git a/actor/lib/worker.js b/actor/lib/worker.js
index ac3cf82..2f5c0c0 100644
--- a/actor/lib/worker.js
+++ b/actor/lib/worker.js
@@ -16,9 +16,11 @@ var mod_assert = require('assert-plus');
 var mod_vasync = require('vasync');
 var mod_once = require('once');
 var mod_verror = require('verror');
+var mod_jsprim = require('jsprim');
 
 var lib_logsets = require('./logsets');
 var lib_findstream = require('./findstream');
+var lib_listglob = require('./listglob');
 var lib_remember = require('./remember');
 
 var VError = mod_verror.VError;
@@ -110,19 +112,63 @@ run(callback)
 	mod_assert.func(callback, 'callback');
 	self.lsw_end_callback = callback;
 
-	/*
-	 * Add zoneroot prefix to search directories if we're looking
-	 * at the non-global zone.
-	 */
-	var adjdirs = [];
-	for (var i = 0; i < self.lsw_logset.search_dirs.length; i++) {
-		adjdirs.push(mod_path.join(self.lsw_path_prefix,
-		    self.lsw_logset.search_dirs[i]));
-	}
+	self.lsw_run_pipeline = mod_vasync.pipeline({
+		arg: self.lsw_manta_task,
+		funcs: [
+			function (_, next) {
+                /*
+                 * If we're not using a search_pattern, skip this part.
+                 */
+                if (!self.lsw_logset.hasOwnProperty('search_pattern')) {
+                    next();
+                    return;
+                }
+                lib_listglob.listGlob(self.lsw_logset.search_pattern,
+                function (err, matches) {
+					if (err) {
+						next();
+						return;
+					}
+
+                    for (var match in matches) {
+                        if (self.lsw_logset.search_dirs
+                                .indexOf(match) === -1)
+                        {
+                            self.lsw_logset.search_dirs =
+                                self.lsw_logset.search_dirs.concat(matches);
+                            }
+                        }
+
+                    next();
+				});
+			}
+		]
+	}, function (err) {
+		if (err) {
+			callback(err);
+			return;
+		}
 
-	self.lsw_barrier.start('find');
-	var find = new lib_findstream.FindStream(adjdirs);
+        /*
+         * Add zoneroot prefix to search directories if we're
+         * looking at the non-global zone.
+         */
+        var adjdirs = [];
+        for (var i = 0; i < self.lsw_logset.search_dirs.length; i++) {
+            adjdirs.push(mod_path.join(self.lsw_path_prefix,
+                self.lsw_logset.search_dirs[i]));
+        }
+		self.lsw_barrier.start('find');
+		self._find(adjdirs);
+	});
+};
 
+
+LogsetWorker.prototype._find = function
+_find(adjdirs)
+{
+    var self = this;
+	var find = new lib_findstream.FindStream(adjdirs);
 	find.on('error', function (err) {
 		/*
 		 * XXX It's possible that there's a class of error we can
@@ -250,6 +296,17 @@ _disp()
 	var delete_after = archive_after + (self.lsw_logset.retain_time * 1000);
 	var _delete = (now >= delete_after);
 
+	/*
+	 * We want to upload files by default, unless the logset specifies
+	 * no_upload as being true.
+	 */
+	var _upload = true;
+	if (self.lsw_logset.no_upload === true) {
+		self.lsw_log.info({ file: inf },
+		    'skipping upload for file because \'no_upload\' specified');
+		_upload = false;
+	}
+
 	/*
 	 * Derive the path that we would upload this file to, in Manta:
 	 */
@@ -273,10 +330,11 @@ _disp()
 	self.lsw_log.debug({
 		file: inf,
 		manta_path: manta_path,
-		do_delete: _delete
+		do_delete: _delete,
+		do_upload: _upload
 	}, 'archiving log file');
 
-	self._manta_upload(inf, manta_path, _delete, function (err) {
+	self._manta_upload(inf, manta_path, _delete, _upload, function (err) {
 		if (err) {
 			self.lsw_log.error({
 				err: err
@@ -305,7 +363,7 @@ _disp()
 function
 pl_local_md5(t, next)
 {
-	if (t.t_cancel) {
+	if (t.t_cancel || !t.t_do_upload) {
 		next();
 		return;
 	}
@@ -333,7 +391,7 @@ pl_local_md5(t, next)
 function
 pl_manta_info(t, next)
 {
-	if (t.t_cancel) {
+	if (t.t_cancel || !t.t_do_upload) {
 		next();
 		return;
 	}
@@ -356,7 +414,7 @@ pl_manta_info(t, next)
 function
 pl_manta_mkdirp(t, next)
 {
-	if (t.t_cancel) {
+	if (t.t_cancel || !t.t_do_upload) {
 		next();
 		return;
 	}
@@ -380,7 +438,7 @@ pl_manta_mkdirp(t, next)
 function
 pl_manta_put(t, next)
 {
-	if (t.t_cancel) {
+	if (t.t_cancel || !t.t_do_upload) {
 		next();
 		return;
 	}
@@ -389,6 +447,7 @@ pl_manta_put(t, next)
 
 	mod_assert.string(t.t_md5_local, 'md5_local');
 	mod_assert.number(t.t_file.size, 'file.size');
+	mod_assert.strictEqual(t.t_do_upload, true);
 
 	var opts = {
 		md5: t.t_md5_local,
@@ -449,7 +508,7 @@ pl_manta_put(t, next)
 function
 pl_compare_hash(t, next)
 {
-	if (t.t_cancel) {
+	if (t.t_cancel || !t.t_do_upload) {
 		next();
 		return;
 	}
@@ -478,9 +537,11 @@ pl_local_rm(t, next)
 		return;
 	}
 
-	mod_assert.string(t.t_md5_remote);
-	mod_assert.string(t.t_md5_local);
-	mod_assert.strictEqual(t.t_md5_remote, t.t_md5_local);
+    if (t.t_do_upload) {
+        mod_assert.string(t.t_md5_remote);
+        mod_assert.strictEqual(t.t_md5_remote, t.t_md5_local);
+        mod_assert.string(t.t_md5_local);
+    }
 	mod_assert.strictEqual(t.t_do_delete, true);
 
 	mod_fs.unlink(t.t_file.real_path, function (err) {
@@ -504,7 +565,7 @@ pl_local_rm(t, next)
 }
 
 LogsetWorker.prototype._manta_upload = function
-_manta_upload(file, manta_path, _delete, next)
+_manta_upload(file, manta_path, _delete, _upload, next)
 {
 	var self = this;
 
@@ -519,6 +580,7 @@ _manta_upload(file, manta_path, _delete, next)
 		t_manta_path: manta_path,
 		t_file: file,
 		t_do_delete: _delete,
+		t_do_upload: _upload,
 		t_md5_remote: null,
 		t_md5_local: null,
 		t_cancel: self.lsw_cancel
diff --git a/lib/logsets.js b/lib/logsets.js
index adf20ad..2752b4e 100644
--- a/lib/logsets.js
+++ b/lib/logsets.js
@@ -28,7 +28,9 @@ load_logsets()
 
 		ls.regex = new RegExp(ls.regex);
 
+		mod_assert.optionalString(ls.search_pattern, 'search_pattern');
 		mod_assert.arrayOfString(ls.search_dirs, 'search_dirs');
+		mod_assert.optionalBool(ls.no_upload, 'no_upload');
 		mod_assert.string(ls.manta_path, 'manta_path');
 
 		mod_assert.optionalNumber(ls.debounce_time, 'debounce_time');
@@ -47,11 +49,13 @@ load_logsets()
 
 var COPY_FIELDS = [
 	'search_dirs',
+	'search_pattern',
 	'manta_path',
 	'date_string',
 	'date_adjustment',
 	'debounce_time',
-	'retain_time'
+	'retain_time',
+	'no_upload'
 ];
 
 /*
